### ISSUE TITLE
Controller pod has been restarted 90 times in the last week

### ISSUE THREAD

**Issue by bourquep** [runconduit/conduit#607]
On my prod cluster, which has been running Conduit 0.3.1 for a week now:

`k get pod -n conduit`

```
NAME                          READY     STATUS    RESTARTS   AGE
controller-598b867959-5qw44   6/6       Running   90         6d
grafana-758656bbf4-x7qhn      2/2       Running   0          6d
prometheus-55b86d854b-wjm7f   2/2       Running   0          6d
web-698bd69459-2b25z          2/2       Running   0          1m
```

Notice the 90 restarts of the controller pod.

---

`k describe pod -n conduit controller-598b867959-5qw44`

```
Name:           controller-598b867959-5qw44
Namespace:      conduit
Node:           gke-studyo-prod-default-pool-6cc404b0-zr5h/10.142.0.4
Start Time:     Fri, 16 Mar 2018 14:01:09 -0400
Labels:         conduit.io/control-plane-component=controller
                conduit.io/control-plane-ns=conduit
                pod-template-hash=1546423515
Annotations:    conduit.io/created-by=conduit/cli v0.3.1
                conduit.io/proxy-version=v0.3.1
                kubernetes.io/created-by={"kind":"SerializedReference","apiVersion":"v1","reference":{"kind":"ReplicaSet","namespace":"conduit","name":"controller-598b867959","uid":"018241b3-2944-11e8-8aed-42010af000...
Status:         Running
IP:             10.32.0.215
Created By:     ReplicaSet/controller-598b867959
Controlled By:  ReplicaSet/controller-598b867959
Init Containers:
  conduit-init:
    Container ID:  docker://5655baddb2d72b3f84562d0d61c5064240c09316c21f3a579953763127e62364
    Image:         gcr.io/runconduit/proxy-init:v0.3.1
    Image ID:      docker-pullable://gcr.io/runconduit/proxy-init@sha256:b82bec4add084a618fe75e419401190ce3044a23706aa846b6f5e59c8225586d
    Port:          <none>
    Args:
      --incoming-proxy-port
      4143
      --outgoing-proxy-port
      4140
      --proxy-uid
      2102
      --inbound-ports-to-ignore
      4190
    State:          Terminated
      Reason:       Completed
      Exit Code:    0
      Started:      Fri, 16 Mar 2018 14:01:11 -0400
      Finished:     Fri, 16 Mar 2018 14:01:11 -0400
    Ready:          True
    Restart Count:  0
    Environment:    <none>
    Mounts:
      /var/run/secrets/kubernetes.io/serviceaccount from conduit-controller-token-xpjlv (ro)
Containers:
  public-api:
    Container ID:  docker://645c669882b7690f1097fd43579400293339c1dcd67a8a73a3cabb59cabcd819
    Image:         gcr.io/runconduit/controller:v0.3.1
    Image ID:      docker-pullable://gcr.io/runconduit/controller@sha256:ad98feaff4d25cc7ef6d70aa6c9a65e4334896363cfb9095fb361587e6e013a2
    Ports:         8085/TCP, 9995/TCP
    Args:
      public-api
      -controller-namespace=conduit
      -log-level=info
      -logtostderr=true
    State:          Running
      Started:      Fri, 16 Mar 2018 14:01:16 -0400
    Ready:          True
    Restart Count:  0
    Environment:    <none>
    Mounts:
      /var/run/secrets/kubernetes.io/serviceaccount from conduit-controller-token-xpjlv (ro)
  destination:
    Container ID:  docker://82bcaa31fb1dc74e60f5bb4a03cedd73559ca176146aca60237ebf232808af7b
    Image:         gcr.io/runconduit/controller:v0.3.1
    Image ID:      docker-pullable://gcr.io/runconduit/controller@sha256:ad98feaff4d25cc7ef6d70aa6c9a65e4334896363cfb9095fb361587e6e013a2
    Ports:         8089/TCP, 9999/TCP
    Args:
      destination
      -log-level=info
      -logtostderr=true
    State:          Running
      Started:      Fri, 16 Mar 2018 14:01:17 -0400
    Ready:          True
    Restart Count:  0
    Environment:    <none>
    Mounts:
      /var/run/secrets/kubernetes.io/serviceaccount from conduit-controller-token-xpjlv (ro)
  proxy-api:
    Container ID:  docker://a7ebbf3b120e23798483b6747c99746cfd2aebf9a4ab7f5493a896607452850c
    Image:         gcr.io/runconduit/controller:v0.3.1
    Image ID:      docker-pullable://gcr.io/runconduit/controller@sha256:ad98feaff4d25cc7ef6d70aa6c9a65e4334896363cfb9095fb361587e6e013a2
    Ports:         8086/TCP, 9996/TCP
    Args:
      proxy-api
      -log-level=info
      -logtostderr=true
    State:          Running
      Started:      Fri, 16 Mar 2018 14:01:17 -0400
    Ready:          True
    Restart Count:  0
    Environment:    <none>
    Mounts:
      /var/run/secrets/kubernetes.io/serviceaccount from conduit-controller-token-xpjlv (ro)
  tap:
    Container ID:  docker://8690ab373e84fc0497953daac78b550fbbf2aeea0874b06b34244e3d3575bd19
    Image:         gcr.io/runconduit/controller:v0.3.1
    Image ID:      docker-pullable://gcr.io/runconduit/controller@sha256:ad98feaff4d25cc7ef6d70aa6c9a65e4334896363cfb9095fb361587e6e013a2
    Ports:         8088/TCP, 9998/TCP
    Args:
      tap
      -log-level=info
      -logtostderr=true
    State:          Running
      Started:      Fri, 16 Mar 2018 14:01:18 -0400
    Ready:          True
    Restart Count:  0
    Environment:    <none>
    Mounts:
      /var/run/secrets/kubernetes.io/serviceaccount from conduit-controller-token-xpjlv (ro)
  telemetry:
    Container ID:  docker://e6030099a2f5e95e00c0b323982dbe79ee9504522aaa481fef6596c207e44819
    Image:         gcr.io/runconduit/controller:v0.3.1
    Image ID:      docker-pullable://gcr.io/runconduit/controller@sha256:ad98feaff4d25cc7ef6d70aa6c9a65e4334896363cfb9095fb361587e6e013a2
    Ports:         8087/TCP, 9997/TCP
    Args:
      telemetry
      -ignore-namespaces=kube-system
      -prometheus-url=http://prometheus.conduit.svc.cluster.local:9090
      -log-level=info
      -logtostderr=true
    State:          Running
      Started:      Fri, 16 Mar 2018 14:01:18 -0400
    Ready:          True
    Restart Count:  0
    Environment:    <none>
    Mounts:
      /var/run/secrets/kubernetes.io/serviceaccount from conduit-controller-token-xpjlv (ro)
  conduit-proxy:
    Container ID:   docker://93201f10b729767bf5f277bbb1fabb6052b2f015682fe15d4ecf755e499dc260
    Image:          gcr.io/runconduit/proxy:v0.3.1
    Image ID:       docker-pullable://gcr.io/runconduit/proxy@sha256:847986d67b64fd68b0c369a9db84dc4f76a661248be61f5fc0b8e877cda42591
    Port:           4143/TCP
    State:          Running
      Started:      Fri, 23 Mar 2018 13:07:16 -0400
    Last State:     Terminated
      Reason:       Error
      Exit Code:    101
      Started:      Fri, 23 Mar 2018 12:57:41 -0400
      Finished:     Fri, 23 Mar 2018 13:07:01 -0400
    Ready:          True
    Restart Count:  90
    Environment:
      CONDUIT_PROXY_LOG:               warn,conduit_proxy=info
      CONDUIT_PROXY_CONTROL_URL:       tcp://localhost:8086
      CONDUIT_PROXY_CONTROL_LISTENER:  tcp://0.0.0.0:4190
      CONDUIT_PROXY_PRIVATE_LISTENER:  tcp://127.0.0.1:4140
      CONDUIT_PROXY_PUBLIC_LISTENER:   tcp://0.0.0.0:4143
      CONDUIT_PROXY_NODE_NAME:          (v1:spec.nodeName)
      CONDUIT_PROXY_POD_NAME:          controller-598b867959-5qw44 (v1:metadata.name)
      CONDUIT_PROXY_POD_NAMESPACE:     conduit (v1:metadata.namespace)
    Mounts:
      /var/run/secrets/kubernetes.io/serviceaccount from conduit-controller-token-xpjlv (ro)
Conditions:
  Type           Status
  Initialized    True 
  Ready          True 
  PodScheduled   True 
Volumes:
  conduit-controller-token-xpjlv:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  conduit-controller-token-xpjlv
    Optional:    false
QoS Class:       BestEffort
Node-Selectors:  <none>
Tolerations:     node.alpha.kubernetes.io/notReady:NoExecute for 300s
                 node.alpha.kubernetes.io/unreachable:NoExecute for 300s
Events:
  Type     Reason      Age                From                                                 Message
  ----     ------      ----               ----                                                 -------
  Warning  BackOff     41m (x7 over 6d)   kubelet, gke-studyo-prod-default-pool-6cc404b0-zr5h  Back-off restarting failed container
  Warning  FailedSync  41m (x7 over 6d)   kubelet, gke-studyo-prod-default-pool-6cc404b0-zr5h  Error syncing pod
  Normal   Created     41m (x91 over 6d)  kubelet, gke-studyo-prod-default-pool-6cc404b0-zr5h  Created container
  Normal   Pulled      41m (x90 over 6d)  kubelet, gke-studyo-prod-default-pool-6cc404b0-zr5h  Container image "gcr.io/runconduit/proxy:v0.3.1" already present on machine
  Normal   Started     41m (x91 over 6d)  kubelet, gke-studyo-prod-default-pool-6cc404b0-zr5h  Started container
```

---

`k get events  --sort-by='.metadata.creationTimestamp' -n conduit`

```
LAST SEEN   FIRST SEEN   COUNT     NAME                                           KIND         SUBOBJECT                           TYPE      REASON                  SOURCE                                                MESSAGE
45m         6d           91        controller-598b867959-5qw44.151c7883611b3dea   Pod          spec.containers{conduit-proxy}      Normal    Started                 kubelet, gke-studyo-prod-default-pool-6cc404b0-zr5h   Started container
45m         6d           90        controller-598b867959-5qw44.151c8fe15df4a8f2   Pod          spec.containers{conduit-proxy}      Normal    Pulled                  kubelet, gke-studyo-prod-default-pool-6cc404b0-zr5h   Container image "gcr.io/runconduit/proxy:v0.3.1" already present on machine
45m         6d           91        controller-598b867959-5qw44.151c788357a90359   Pod          spec.containers{conduit-proxy}      Normal    Created                 kubelet, gke-studyo-prod-default-pool-6cc404b0-zr5h   Created container
45m         6d           7         controller-598b867959-5qw44.151cb385400d97d9   Pod                                              Warning   FailedSync              kubelet, gke-studyo-prod-default-pool-6cc404b0-zr5h   Error syncing pod
45m         6d           7         controller-598b867959-5qw44.151cb3853ffa2160   Pod          spec.containers{conduit-proxy}      Warning   BackOff                 kubelet, gke-studyo-prod-default-pool-6cc404b0-zr5h   Back-off restarting failed container
```

---

This does not affect runtime behavior - all calls are being proxied as expected, and live stats are reported by `conduit stats` and `conduit dashboard`.

However, `conduit dashboard` does not reflect deployment scale changes (e.g. scaling up a deployment, dashboard still shows the old cound of green dots).

I won't touch anything for now, if there is any other info you need me to provide I'll be happy to do so.

**Comment by bourquep** [comment]
I have killed the `controller` pod yesterday evening, and it already has 10 restarts.

Here's the relevant subset of `kubectl describe pod controller`:

```
  conduit-proxy:
    Container ID:   docker://692cca7c1ac0b570e0a01a4d24704e2b033cd3694991dc6839871452230e5937
    Image:          gcr.io/runconduit/proxy:v0.3.1
    Image ID:       docker-pullable://gcr.io/runconduit/proxy@sha256:847986d67b64fd68b0c369a9db84dc4f76a661248be61f5fc0b8e877cda42591
    Port:           4143/TCP
    State:          Running
      Started:      Sat, 24 Mar 2018 10:18:45 -0400
    Last State:     Terminated
      Reason:       Error
      Exit Code:    101
      Started:      Sat, 24 Mar 2018 08:49:09 -0400
      Finished:     Sat, 24 Mar 2018 10:18:44 -0400
    Ready:          True
    Restart Count:  10
```

And here are the logs for the `conduit-proxy` container in the pod, for the last start..finish time range (08:40..10:18):

```

2018-03-24 08:49:09.000 EDT
INFO conduit_proxy using controller at HostAndPort { host: DnsName("localhost"), port: 8086 }
2018-03-24 08:49:09.000 EDT
INFO conduit_proxy routing on V4(127.0.0.1:4140)
2018-03-24 08:49:09.000 EDT
INFO conduit_proxy proxying on V4(0.0.0.0:4143) to None
2018-03-24 08:49:09.000 EDT
INFO conduit_proxy::transport::connect "controller-client", DNS resolved DnsName("localhost") to 127.0.0.1
2018-03-24 08:49:09.000 EDT
thread 'main' panicked at 'must be scheduled to reset', /checkout/src/libcore/option.rs:891:5
2018-03-24 08:49:09.000 EDT
note: Run with `RUST_BACKTRACE=1` for a backtrace.
2018-03-24 10:18:44.000 EDT
thread 'main' panicked at 'must be scheduled to reset', /checkout/src/libcore/option.rs:891:5
2018-03-24 10:18:44.000 EDT
note: Run with `RUST_BACKTRACE=1` for a backtrace.
2018-03-24 10:18:45.000 EDT
INFO conduit_proxy using controller at HostAndPort { host: DnsName("localhost"), port: 8086 }
2018-03-24 10:18:45.000 EDT
INFO conduit_proxy routing on V4(127.0.0.1:4140)
2018-03-24 10:18:45.000 EDT
INFO conduit_proxy proxying on V4(0.0.0.0:4143) to None
2018-03-24 10:18:45.000 EDT
INFO conduit_proxy::transport::connect "controller-client", DNS resolved DnsName("localhost") to 127.0.0.1
 ```

**Comment by hawkw** [comment]
Interesting, looks like the controller's proxy container is hitting [a panic in`h2`](https://github.com/carllerche/h2/blob/f8baeb7211985834acaf7ecc70548aafd5d9ef6d/src/proto/streams/prioritize.rs#L693-L694). 

I think more verbose proxy logs (in particular, setting `h2=trace`) could be helpful for debugging this.

**Comment by bourquep** [comment]
Here's the proxy log with `h2=trace` prior to the last panic that occured.

[h2_trace.log](https://github.com/runconduit/conduit/files/1847857/h2_trace.log)

**Comment by bourquep** [comment]
And here's the stack backtrace (after setting `RUST_BACKTRACE=1` in env):

```
E  stack backtrace:
 
E     0: std::sys::unix::backtrace::tracing::imp::unwind_backtrace
 
E               at ./checkout/src/libstd/sys/unix/backtrace/tracing/gcc_s.rs:49
 
E     1: std::sys_common::backtrace::print
 
E               at ./checkout/src/libstd/sys_common/backtrace.rs:68
 
E               at ./checkout/src/libstd/sys_common/backtrace.rs:57
 
E     2: std::panicking::default_hook::{{closure}}
 
E               at ./checkout/src/libstd/panicking.rs:381
 
E     3: std::panicking::default_hook
 
E               at ./checkout/src/libstd/panicking.rs:397
 
E     4: std::panicking::rust_panic_with_hook
 
E               at ./checkout/src/libstd/panicking.rs:577
 
E     5: std::panicking::begin_panic
 
E               at ./checkout/src/libstd/panicking.rs:538
 
E     6: std::panicking::begin_panic_fmt
 
E               at ./checkout/src/libstd/panicking.rs:522
 
E     7: rust_begin_unwind
 
E               at ./checkout/src/libstd/panicking.rs:498
 
E     8: core::panicking::panic_fmt
 
E               at ./checkout/src/libcore/panicking.rs:71
 
E     9: core::option::expect_failed
 
E               at ./checkout/src/libcore/option.rs:891
 
E    10: h2::proto::streams::prioritize::Prioritize::pop_frame
 
E    11: h2::proto::streams::send::Send::poll_complete
 
E    12: <h2::proto::streams::streams::Streams<B, P>>::poll_complete
 
E    13: <h2::proto::connection::Connection<T, P, B>>::poll
 
E    14: <h2::server::Connection<T, B> as futures::stream::Stream>::poll
 
E    15: <tower_h2::server::Connection<T, S, E, B, F> as futures::future::Future>::poll
 
E    16: <futures::future::map_err::MapErr<A, F> as futures::future::Future>::poll
 
E    17: <futures::future::chain::Chain<A, B, C>>::poll
 
E    18: futures::task_impl::std::set
 
E    19: <futures::task_impl::Spawn<T>>::poll_future_notify
 
E    20: tokio_core::reactor::Core::poll
 
E    21: tokio_core::reactor::Core::run
 
E    22: conduit_proxy::main
 
E    23: std::rt::lang_start::{{closure}}
 
E    24: std::panicking::try::do_call
 
E               at ./checkout/src/libstd/rt.rs:59
 
E               at ./checkout/src/libstd/panicking.rs:480
 
E    25: __rust_maybe_catch_panic
 
E               at ./checkout/src/libpanic_unwind/lib.rs:101
 
E    26: std::rt::lang_start_internal
 
E               at ./checkout/src/libstd/panicking.rs:459
 
E               at ./checkout/src/libstd/panic.rs:365
 
E               at ./checkout/src/libstd/rt.rs:58
 
E    27: main
 
E    28: __libc_start_main
 
E    29: _start
 
```

**Comment by wmorgan** [comment]
Thanks @bourquep!

**Comment by hawkw** [comment]
I suspect @seanmonstar or @carllerche might be able to provide some additional insight into what might cause us to hit that `expect` in `h2`, and whether it's a Conduit bug or a `h2` bug?

**Comment by carllerche** [comment]
It looks like it is hitting this: https://github.com/carllerche/h2/blob/master/src/proto/streams/prioritize.rs#L693

This is definitely a bug in `h2`, though I am not immediately sure what is causing it. Is there any other context that might be available?

I will dig in. I might come back asking for more logs though :)

**Comment by hawkw** [comment]
Hi @bourquep, I've opened carllerche/h2#246 to track the issue that seems to be causing your proxies to crash.

**Comment by hawkw** [comment]
hawkw/h2@7b80ac0 adds a test to `h2` that reproduces the crash.

**Comment by carllerche** [comment]
I have no idea why Github decided to close this.

**Comment by hawkw** [comment]
This issue should be closed by 308f7a47001e5ebd7b19b9c4e7823960dc1f5e1b, correct?

**Comment by carllerche** [comment]
Yes! At least I sure hope so.
